readthedocs/embed/v3/views.py

```diff
@@ -112,10 +113,11 @@ def _get_page_content_from_storage(self, project, version, filename):
         )
 
         try:
-            with build_media_storage.open(
-                file_path
-            ) as fd:  # pylint: disable=invalid-name
-                return fd.read()
+            tryfiles = [file_path, os.path.join(file_path, "index.html")]
```

Comment:
You should use build_media_storage.join, otherwise this is likely vulnerable to path traversal. Also, I think we already have some logic for handling index files somewhere, we just check if the file ends with / and append index.html, otherwise just treat it as a regular file.

Reply:
I found this code: https://github.com/readthedocs/readthedocs.org/blob/7195740a65dd29dc25c96534c68a0c0e7171417a/readthedocs/proxito/views/mixins.py#L66-L76

Reply:
I updated this PR following that code 👍🏼

Reply:
> we just check if the file ends with / and append index.html

We can't do that in this case. The `filename` I'm receiving when using Docusaurus doesn't have an ending `/`:

```
(Pdb++) filename
'/docs/tutorial-basics/create-a-page'
```

I'm not finding _where_ in the code we manage this case from El Proxito. These type of URLs are definitely being served correctly.

Reply:
I found _why_ it works on El Proxito, we are performing a redirect when the URL doesn't ends with `/`

```
▶ curl -sIL http://test-builds.devthedocs.org/en/docusaurus/docs/tutorial-basics/create-a-page | head
HTTP/1.1 302 Found
Server: nginx/1.22.0
Date: Tue, 03 Dec 2024 07:41:12 GMT
Content-Type: text/html; charset=utf-8
Content-Length: 0
Connection: keep-alive
Location: /en/docusaurus/docs/tutorial-basics/create-a-page/
Access-Control-Allow-Origin: *
Content-Language: en
Vary: Accept-Language, Cookie
...
```

Reply:
yeah, we always redirect to the correct path instead of trying to serve the file from two different URLs

Reply:
> Do you mean that Docusaurus creates links without /?

Yes.

There are a few documentation tools that don't use ending `/`, in particular those that are SPA for some reason.

If you go to https://test-builds.readthedocs.io/en/docusaurus/docs/intro we will redirect you to https://test-builds.readthedocs.io/en/docusaurus/docs/intro/. However, if you click any link, the resulting URL in your browser won't have that ending `/` -- that's the URL we are sending to the backend, and since we don't find the file in the storage we are returning 404.

Checking for the `index.html` solves this issue.